### PR TITLE
protobuf: add extension to support extended protobuf modes

### PR DESCRIPTION
**Pull request by rohanjoyce**
There are a number of internal-only protobuf modes/flags that have dormant code
left in the open source protobuf plugin. This change allows these modes/flags
to be implemented via an extension point and simplifies the associated code in
the open source plugin.

Summary: 
  - new extension point `com.intellij.protobuf.nameGeneratorContributor` for this purpose, associated interface is in `NameGeneratorContributer.java`
  - removal of support for Proto1/Proto2Mutable, and the flags associated with those (not in open source protobufs)
  - widen visibility of some existing methods/classes to allow implementations of the above extension point to use them

### CODE REVIEW

File: protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  static Optional<String> parseStringOption(PbOptionExpression optionExpression) {
+    return Optional.ofNullable(optionExpression.getStringValue())
+      .map(ProtoStringValue::getAsString);
+  }
+
+  static Optional<Boolean> parseBoolOption(PbOptionExpression optionExpression) {
+    return Optional.ofNullable(optionExpression.getIdentifierValue())
+      .map(PbIdentifierValue::getBooleanValue);
+  }
+}
```
Comment (NikitaKatkov) [inline protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java:61]: Should we extract these utility methods to some new/existing `*Utils` class? For me it looks better when EP interface does contain only methods intended to be implemented

**Review by NikitaKatkov — CHANGES_REQUESTED** [review]

File: protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+   *
+   * @param file the protobuf file to provide name generators for
+   * @return boolean indicating whether or not the extension will process the given file
+   */
+  boolean isApplicable(PbFile file);
+
+  /**
+   * Provides name generators for fields contained in the given protobuf file.
+   *
+   * <p><b>Warning:</b> This method should be called in a read action compatible context.
```
[the remark this answers is not in the document]
Comment (rohanjoyce) [reply protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java]: Added RequiresReadLock, thanks!

File: protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java (hunk 1 above)
Comment (rohanjoyce) [reply protobuf/protoeditor-jvm/src/com/intellij/protobuf/jvm/names/NameGeneratorContributor.java]: Extracted these out to com.intellij.protobuf.lang.names.NameGeneratorUtils

File: protobuf/protoeditor-jvm/resources/intellij.protoeditor.jvm.xml [hunk 3]
```diff
@@ -13,4 +13,9 @@
     <findUsagesHandlerFactory
       implementation="com.intellij.protobuf.jvm.PbJavaFindUsagesHandlerFactory"/>
   </extensions>
+  <extensionPoints>
+    <extensionPoint
```
[the remark this answers is not in the document]
Comment (rohanjoyce) [reply protobuf/protoeditor-jvm/resources/intellij.protoeditor.jvm.xml]: Lifted the declaration to the main plugin.xml with dynamic="true", and shifted the interface to protoeditor-core. Some minor changes were made to the interface to break a Java-specific dependency.

**Comment by NikitaKatkov** [comment]
thanks a lot for the provided fixes. merged to master

**Pull request closed without merging**